### ISSUE TITLE
Provide passthrough `Code` property for `AWS::Serverless::Function`.

### ISSUE THREAD

**Issue by garretwilson** [aws/serverless-application-model#3358]
SAM's [`AWS::Serverless::Function`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html) removes the ability to use imported values in referring to the bucket and filename of my uploaded ZIP file for AWS Lambda. This is an inherent issue with the SAM transform design; see e.g. #2533 and a long [comment in another ticket](https://github.com/aws/aws-sam-cli/issues/5249#issuecomment-1574549344) explaining all the gory details.

As explained in detail in #3264, I can do the following just fine using [AWS::Lambda::Function](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html), which uses [Code](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-code) instead of [CodeUri](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html#sam-function-codeuri). It looks like this:

```yaml
      Code:
        S3Bucket:
          Fn::ImportValue:
            !Sub "other-stack-${env}:stagingBucket"
        S3Key: !Sub "foo-${ver}-aws-lambda.zip"
```

But I can't do that with [`AWS::Serverless::Function`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html):

```yaml
      CodeUri: !Sub
        - "s3://${bucket}/foo-${ver}-aws-lambda.zip"
        - bucket: my-staging-bucket
            Fn::ImportValue:
              !Sub "other-stack-${env}:stagingBucket"
```

That produces:

> `Error: Failed to create changeset for the stack: my-stack, ex: Waiter ChangeSetCreateComplete failed: Waiter encountered a terminal failure state: For expression "Status" we matched expected path: "FAILED" Status: FAILED. Reason: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [MyFunction] is invalid. 'CodeUri' requires Bucket and Key properties to be specified.`

Please add a [Code](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-code) pass-through property to [AWS::Serverless::Function](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html) so we can at least work around this shortcoming in the design of SAM.

**Comment by GavinZZ** [comment]
Hi thanks for creating a issue. I apologize for the trouble you've been through. I'll bring it up for discussion with the team and get back to you soon.

**Comment by GavinZZ** [comment]
Hi @garretwilson, I've brought this up to the team and we're having discussion on potential workaround/solution. Meanwhile I just want to quickly confirm that you've also tried with the following syntax and it doesn't work for you.
```
  CodeUri:
    Bucket:
      Fn::ImportValue:
        !Sub "other-stack-${env}:stagingBucket"
    Key: !Sub "foo-${ver}-aws-lambda.zip"
```

**Comment by garretwilson** [comment]
Well I'd like to answer your question, but suddenly SAM is doing all sorts of crazy things, as I just reported in #3361. (Sigh. If it's not one thing, it's another thing breaking.)

**Comment by garretwilson** [comment]
@GavinZZ , I have wonderful news! (Note that the SAM traceback in #3361 is still happening, but it doesn't seem to impede the actual deployment.) The format you asked about worked:

```yaml
  CodeUri:
    Bucket:
      Fn::ImportValue:
        !Sub "other-stack-${env}:stagingBucket"
    Key: !Sub "foo-${ver}-aws-lambda.zip"
```

It seems that [`CodeUri`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html#sam-function-codeuri) allows not only a `String` but also a [`FunctionCode`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-functioncode.html) object, the latter of which has a `Bucket` and a `Key`. These seems to be passthrough properties for CloudFormation `S3Bucket` and `S3Key`. Yay! 🎉 Just what I need.

I wonder why nobody mentioned this during the log discussion in #3264? 🤔 

Is the [`CodeUri` > `FunctionCode`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-functioncode.html) type new for SAM, or has it always been there?

**Comment by GavinZZ** [comment]
Great to hear that and hopefully this syntax will unblock you. 🎉🎉🎉

`CodeUri` supported both `String` syntax and `FunctionCode` object syntax and it's always been this way.

**Comment by garretwilson** [comment]
> 'CodeUri' supported both 'String' syntax and 'FunctionCode' object syntax and it's always been this way.

In that case, feel free to close this current ticket as "invalid", because SAM already has exactly what I was requesting!

Note also that #3264 therefore doesn't impact me, since I can always use the passthrough properties.

I'm so happy to find this out. Thanks for taking the time to suggest that.
